crates/zkvm/zkvm/src/lib.rs

```diff
@@ -204,3 +204,102 @@ impl AggregationInput {
         &self.vk
     }
 }
+
+/// A trait representing a Zero-Knowledge Virtual Machine (ZkVM) interface.
+/// Provides methods for reading inputs, committing outputs, and verifying proofs
+/// within the ZkVM environment.
+pub trait ZkVm {
+    /// Reads a serialized object from the guest code, deserializing it using Serde.
+    ///
+    /// The input is expected to be written with [`ZKVMInputBuilder::write`].
+    fn read<T: DeserializeOwned>(&self) -> T;
+
+    /// Reads a serialized byte slice from the guest code.
+    ///
+    /// The input is expected to be written with [`ZKVMInputBuilder::write_serialized`].
+    fn read_slice(&self) -> Vec<u8>;
+
+    /// Commits a Serde-serializable object to the public values stream.
+    ///
+    /// Values that are committed can be proven as public parameters.
+    fn commit<T: Serialize>(&self, output: &T);
+
+    /// Commits a pre-serialized byte slice to the public values stream.
+    ///
+    /// This method is intended for cases where the data has already been serialized
+    /// outside of the ZkVM's standard serialization methods. It allows you to provide
+    /// serialized outputs directly, bypassing any further serialization.
+    fn commit_slice(&self, raw_output: &[u8]);
+
+    /// Verifies a proof generated with the ZkVM.
+    ///
+    /// This method checks the validity of the proof against the provided verification key digest
+    /// and public values. It will panic if the proof fails to verify.
+    fn verify_proof(&self, vk_digest: &[u32; 8], public_values: &[u8]);
+
+    /// Verifies a Groth16 proof.
+    ///
+    /// # Parameters
+    ///
+    /// * `proof`: A byte slice containing the serialized proof. (TODO: Change `proof: &[u8]` to a
+    ///   `Proof` type.)
+    /// * `verification_key`: A byte slice containing the serialized verification key.
+    /// * `public_params_raw`: A byte slice containing the serialized public parameters.
+    ///
+    /// # Returns
+    ///
+    /// Returns `Ok(())` if the proof verifies successfully, or an `anyhow::Error` if verification
+    /// fails.
+    fn verify_groth16(
+        &self,
+        proof: &[u8],
+        verification_key: &[u8],
+        public_params_raw: &[u8],
+    ) -> anyhow::Result<()>;
```

Comment:
First maybe should be `verify_native_proof`?  And then these I would put after the io fns to keep them together.